File: packages/engine/src/schedule.test.ts
import { describe, expect, it } from "vitest";
import { resolveScheduleConfig, scheduleExecutor, validateCronExpression } from "./schedule";

describe("validateCronExpression", () => {
  it("accepts standard 5-field cron expressions and returns a next-fire ISO string", () => {
    const result = validateCronExpression("0 9 * * *");
    expect(result.nextFireAt).toMatch(/^\d{4}-\d{2}-\d{2}T/);
  });

  it("accepts every-N-minutes expressions", () => {
    expect(() => validateCronExpression("*/5 * * * *")).not.toThrow();
    expect(() => validateCronExpression("0 */6 * * *")).not.toThrow();
  });

  it("rejects empty / missing cron expressions with a descriptive error", () => {
    expect(() => validateCronExpression("")).toThrow(/cronExpression is required/);
    expect(() => validateCronExpression("   ")).toThrow(/cronExpression is required/);
    expect(() => validateCronExpression(undefined)).toThrow(/cronExpression is required/);
    expect(() => validateCronExpression(null)).toThrow(/cronExpression is required/);
    expect(() => validateCronExpression(42)).toThrow(/cronExpression is required/);
  });

  it("rejects malformed cron strings", () => {
    expect(() => validateCronExpression("99 9 * * *")).toThrow(/valid cron expression/);
    expect(() => validateCronExpression("0 25 * * *")).toThrow(/valid cron expression/);
  });

  it("rejects non-5-field cron expressions so schedules stay minute-granular", () => {
    expect(() => validateCronExpression("not a cron")).toThrow(/5-field cron expression/);
    expect(() => validateCronExpression("0 0 9 * * *")).toThrow(/5-field cron expression/);
  });
});

describe("resolveScheduleConfig", () => {
  it("returns a normalized config when cron is valid and enabled defaults to true", () => {
    expect(resolveScheduleConfig({ cronExpression: "0 9 * * *" })).toEqual({
      cronExpression: "0 9 * * *",
      enabled: true,
    });
  });

  it("respects an explicit enabled=false", () => {
    expect(resolveScheduleConfig({ cronExpression: "0 9 * * *", enabled: false })).toEqual({
      cronExpression: "0 9 * * *",
      enabled: false,
    });
  });

  it("trims surrounding whitespace from cronExpression", () => {
    expect(resolveScheduleConfig({ cronExpression: "  0 9 * * *  " }).cronExpression).toBe("0 9 * * *");
  });

  it("rejects non-object config", () => {
    expect(() => resolveScheduleConfig(null)).toThrow(/config object/);
    expect(() => resolveScheduleConfig([])).toThrow(/config object/);
    expect(() => resolveScheduleConfig("nope")).toThrow(/config object/);
  });

  it("rejects non-boolean enabled when set", () => {
    expect(() => resolveScheduleConfig({ cronExpression: "0 9 * * *", enabled: "yes" })).toThrow(/enabled must be a boolean/);
    expect(() => resolveScheduleConfig({ cronExpression: "0 9 * * *", enabled: 1 })).toThrow(/enabled must be a boolean/);
  });

  it("propagates cron validation errors", () => {
    expect(() => resolveScheduleConfig({ cronExpression: "bogus" })).toThrow(/5-field cron expression/);
    expect(() => resolveScheduleConfig({})).toThrow(/cronExpression is required/);
  });
});

describe("scheduleExecutor", () => {
  it("succeeds immediately with triggeredAt and the resolved cron echoed back", async () => {
    const result = await scheduleExecutor({
      runId: "r1",
      nodeId: "s1",
      orgId: "default",
      workflowId: "w1",
      config: { cronExpression: "0 9 * * *", enabled: true },
      context: {},
    });
    expect(result.status).toBe("completed");
    if (result.status === "completed") {
      expect(result.output?.cronExpression).toBe("0 9 * * *");
      expect(typeof result.output?.triggeredAt).toBe("string");
      expect(String(result.output?.triggeredAt)).toMatch(/^\d{4}-\d{2}-\d{2}T/);
    }
  });

  it("still succeeds when enabled=false (pause is enforced at the BullMQ register layer, not in the executor)", async () => {
    const result = await scheduleExecutor({
      runId: "r1",
      nodeId: "s1",
      orgId: "default",
      workflowId: "w1",
      config: { cronExpression: "0 9 * * *", enabled: false },
      context: {},
    });
    expect(result.status).toBe("completed");
  });

  it("fails fast on invalid config", async () => {
    await expect(
      scheduleExecutor({
        runId: "r1",
        nodeId: "s1",
        orgId: "default",
        workflowId: "w1",
        config: { cronExpression: "" },
        context: {},
      }),
    ).rejects.toThrow(/cronExpression is required/);
  });
});

File: internal/domain/schedule_test.go
package domain

import (
	"strings"
	"testing"
)

func TestResolveScheduleConfig(t *testing.T) {
	resolved, err := ResolveScheduleConfig(map[string]any{
		"cronExpression": " 0 9 1 * * ", "enabled": false,
	})
	if err != nil || resolved.CronExpression != "0 9 1 * *" || resolved.Enabled {
		t.Fatalf("resolved schedule: %+v err=%v", resolved, err)
	}

	cases := []struct {
		name    string
		config  map[string]any
		message string
		exact   bool
	}{
		{"legacy field", map[string]any{"cron": "0 9 * * *"}, "cronExpression is required", false},
		{"wrong field count", map[string]any{"cronExpression": "0 9 * *"}, "5-field cron expression", false},
		{"invalid cron", map[string]any{"cronExpression": "60 9 * * *"}, "valid cron expression", false},
		{"invalid string enabled", map[string]any{"cronExpression": "0 9 * * *", "enabled": "yes"}, "schedule.enabled must be a boolean when set, got: string", true},
		{"invalid numeric enabled", map[string]any{"cronExpression": "0 9 * * *", "enabled": float64(1)}, "schedule.enabled must be a boolean when set, got: number", true},
	}
	for _, tc := range cases {
		t.Run(tc.name, func(t *testing.T) {
			_, err := ResolveScheduleConfig(tc.config)
			if err == nil || (tc.exact && err.Error() != tc.message) || (!tc.exact && !strings.Contains(err.Error(), tc.message)) {
				t.Fatalf("got %v, want %q", err, tc.message)
			}
		})
	}
}
